lib/tokens.js

```diff
@@ -0,0 +1,687 @@
+  return token;
+}
+
+async function _parseToken({token}) {
+  // validate token
+  if(!(token instanceof Uint8Array && token.length >= VERSION_SIZE)) {
+    throw new TypeError(
+      `"token" must be a Uint8Array that is ${VERSION_SIZE} bytes ` +
+      'or more in size.');
+  }
+
+  // parse version from token
+  const dv = new DataView(token.buffer, token.byteOffset, token.length);
+  const version = dv.getUint16(0);
+
+  // get token version by `version` ID
+  const {tokenVersion} = await tokenVersions.get({id: version});
+  const {options: {batchIdSize, batchSaltSize}} = tokenVersion;
+
+  // validate token is appropriate size
+  // format: version|salt|wrapped|attributes(aad)
+  // version = 2 bytes
+  // salt = batchSaltSize bytes
+  // wrapped = 32 bytes
+  // attributes size only known once `wrapped` is unwrapped
+  // Note: `version` and `salt` map to crypto used to generate a key and will
+  // therefore be authenticated via key unwrapping. The `attributes` can be
+  // compared against the unwrapped attributes to be authenticated as well
+  const minimumSize = VERSION_SIZE + batchSaltSize + WRAPPED_SIZE;
+  // 8 bytes is maximum size for attributes
+  const maximumSize = minimumSize + MAX_AAD_SIZE;
+  if(token.length < minimumSize || token.length > maximumSize) {
+    throw new BedrockError(
+      'Invalid token.',
+      'DataError', {
+        public: true,
+        httpStatusCode: 400
+      });
+  }
+
+  // parse `salt` and `wrapped` and `attributes`
+  let offset = token.byteOffset + VERSION_SIZE;
+  const salt = new Uint8Array(token.buffer, offset, batchSaltSize);
+  const wrapped = new Uint8Array(
+    token.buffer, offset += batchSaltSize, WRAPPED_SIZE);
+  const attributes = new Uint8Array(
+    token.buffer, offset += WRAPPED_SIZE, token.length - minimumSize);
+  // get tokenizer associated with version
+  const tokenizer = await tokenizers.get({id: tokenVersion.tokenizerId});
+  const {hmac} = tokenizer;
+
+  // create KEK via HMAC(version|salt)
+  const kek = await _getKek({hmac, version, salt});
+
+  // unwrap `wrapped` as if it were a key
+  let unwrapped;
+  try {
+    unwrapped = await kek.unwrapKey({wrappedKey: wrapped});
+    if(unwrapped === null) {
+      throw new Error('Decryption failed.');
+    }
+  } catch(e) {
+    throw new BedrockError(
+      'Invalid token.',
+      'DataError', {
+        public: true,
+        httpStatusCode: 400
+      }, e);
+  }
+
+  // at this point, unwrapped is authenticated and known to be 32 bytes
+  // next, determine if full token is valid
+
+  // parse unwrapped: batchId|index|aadSize|aad|padding
+  offset = unwrapped.byteOffset;
+  const batchId = new Uint8Array(unwrapped.buffer, offset, batchIdSize);
+  const index = unwrapped[batchIdSize];
+  const aadSize = unwrapped[batchIdSize + 1];
+  // time-safe compare `aad` against given attributes
+  // always use length of *given* `attributes` to determine what to compare
+  // and then compare actual lengths thereafter
```

Comment:
This could use some more words.  I think it's trying to avoid leaking the wrapped aad size via a timing attack?  If so then I think the `A && B` check needs to avoid short-circuit timing issues too?  Else you could probe with increasing attributes length, byte by byte, finding each one when it takes slightly longer for timingSafeEqual to pass but length check to fail.  Hard to write code to outsmart the runtime for things like this.  Probably need to compare buffers that have the length value and content concatenated?

Reply:
Yeah, this should be tweaked to ensure the length check always happens.